### PR TITLE
IBX-4642: Storefront config

### PR DESCRIPTION
**Pull request by juskora**
| Question      | Answer
| ------------- | ---
| JIRA Ticket   | https://issues.ibexa.co/browse/IBX-4642
| Versions      | master, 4.4

#### Checklist

- [ ] Text renders correctly
- [ ] Text has been checked with vale
- [ ] Description metadata is up to date
- [ ] Redirects cover removed/moved pages
- [ ] Code samples are working
- [ ] PHP code samples have been fixed with PHP CS fixer
- [ ] Added link to this PR in relevant JIRA ticket or code PR

### CODE REVIEW

File: docs/commerce/storefront/storefront.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+---
+description: Storefront covers all actions related to the purchase process.
+edition: commerce
+---
+
+# Storefront
+
+Storefront package provides a starting kit for the developers.
+It is a set of basic components that serves as a basis, which developers can
```
Comment (dabrt) [inline docs/commerce/storefront/storefront.md:9]:
```suggestion
It is a set of components that serves as a basis, which developers can 
```

File: docs/commerce/storefront/configure_storefront.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
+---
+description: Configure Storefront
```
Comment (DominikaK) [inline docs/commerce/storefront/configure_storefront.md:2]:
```suggestion
description: Configure Storefront, including catalogs used, customer groups and user accounts.
```

File: docs/commerce/storefront/configure_storefront.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+edition: commerce
+---
+# Configure Storefront
+
+The Storefront is accesible under the `<yourdomain>/product-catalog`.
```
Comment (DominikaK) [inline docs/commerce/storefront/configure_storefront.md:7]:
```suggestion
The Storefront is accessible under the `<yourdomain>/product-catalog`.
```

File: docs/commerce/storefront/configure_storefront.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+## Catalog configuration
+
+With the `ibexa/storefront` package, you can configure the product catalog and make it available to your shop users. 
+
+Before you start configuring the Storefront, make sure you have cerated and configured [catalogs](https://doc.ibexa.co/projects/userguide/en/latest/pim/work_with_catalogs/#create-catalogs) in the Back Office.
```
Comment (DominikaK) [inline docs/commerce/storefront/configure_storefront.md:13]:
```suggestion
Before you start configuring the Storefront, make sure you have created and configured [catalogs](https://doc.ibexa.co/projects/userguide/en/latest/pim/work_with_catalogs/#create-catalogs) in the Back Office.
```

File: docs/commerce/storefront/configure_storefront.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+The configuration is avaible under the `ibexa.system.<scope>.storefront.catalog` key.
```
Comment (DominikaK) [inline docs/commerce/storefront/configure_storefront.md:15]:
```suggestion
The configuration is available under the `ibexa.system.<scope>.storefront.catalog` key.
```

File: docs/commerce/storefront/configure_storefront.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+1\. All products available for all users:
+
+```yaml
+ibexa:
+    system:
+        site:
+            storefront:
+                catalog: ~
+```
+If the `null` as a key value is provided, the Storefront makes the main product catalog (with all products) visibile for all users.
```
Comment (DominikaK) [inline docs/commerce/storefront/configure_storefront.md:27]:
```suggestion

If `null`is provided as the value, the Storefront makes the main product catalog (with all products) visible for all users.
```

File: docs/commerce/storefront/configure_storefront.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+2\. To expose a single catalog with an identifier to all users only a catalog, provide a `string` value under the key:
```
Comment (DominikaK) [inline docs/commerce/storefront/configure_storefront.md:29]:
```suggestion
2\. To expose a single catalog with an identifier to all users, provide a string value of the catalog identifier:
```

File: docs/commerce/storefront/configure_storefront.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+ibexa:
+    system:
+        site:
+            storefront:
+                catalog: custom_catalog
+```
+
+3\. Specific catalog for the defined customer group
+
+First, create catalogs in the Back Office.
```
Comment (DominikaK) [inline docs/commerce/storefront/configure_storefront.md:41]:
```suggestion
```
This is repetition of an earlier step

File: docs/commerce/storefront/configure_storefront.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+You can expose different catalogs based on a customer group assigned to the current user.
+
+Next, provide the following configuration:
```
Comment (DominikaK) [inline docs/commerce/storefront/configure_storefront.md:45]:
```suggestion
To do it, provide the following configuration:
```

File: docs/commerce/storefront/configure_storefront.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        site:
+            storefront:
+                catalog:
+                    default: standard
+                    customer_group:
+                        retailer: retailer_catalog
+                        wholesale: wholesaler_catalog
+```
+
+The basic configuration of the Storefront can look as the following:
```
Comment (DominikaK) [inline docs/commerce/storefront/configure_storefront.md:59]:
```suggestion
The basic configuration of the Storefront can look as follows:
```

File: docs/commerce/storefront/configure_storefront.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+``` yaml
+[[= include_file('code_samples/front/shop/storefront/config/packages/ibexa.yaml') =]]
+```
+
+## Retrieve catalog assigned to user
+
+`null` stands for the current user
```
Comment (DominikaK) [inline docs/commerce/storefront/configure_storefront.md:67]:
```suggestion
```

File: docs/commerce/storefront/configure_storefront.md [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+use Ibexa\Contracts\Core\Repository\Values\User\User;
+use Ibexa\Contracts\ProductCatalog\Values\CatalogInterface;
+
+interface CatalogResolverInterface
+{
+    public function resolveCatalog(?User $user = null): ?CatalogInterface;
+}
+```
+
```
Comment (DominikaK) [inline docs/commerce/storefront/configure_storefront.md:83]:
```suggestion
`null` stands for the current user.

```

File: docs/commerce/storefront/configure_storefront.md [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+``` yaml
+[[= include_file('code_samples/front/shop/storefront/config/packages/ibexa.yaml') =]]
+```
+
+## Retrieve catalog assigned to user
+
+`null` stands for the current user
+
+The `\Ibexa\Contracts\Storefront\Repository\CatalogResolverInterface` service allows to retrieve product catalog available for a specific user.
```
Comment (DominikaK) [inline docs/commerce/storefront/configure_storefront.md:69]:
```suggestion
The `\Ibexa\Contracts\Storefront\Repository\CatalogResolverInterface` interface allows retrieving the product catalog available for a specific user.
```

File: docs/commerce/storefront/configure_storefront.md [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+{
+    public function resolveCatalog(?User $user = null): ?CatalogInterface;
+}
+```
+
+### Configure user account
+
+The following user settings mechanisms used in `ibexa/storefront` are reused from `ibexa/user` package:
+
+- [change password feature](../../users/user_management.md)
```
Comment (DominikaK) [inline docs/commerce/storefront/configure_storefront.md:88]:
```suggestion
- [change password feature](user_management.md)
```

File: docs/commerce/storefront/extend_storefront.md [hunk 15]
```diff
@@ -0,0 +1,92 @@
+---
+description: Extend Storefront configuration
```
Comment (DominikaK) [inline docs/commerce/storefront/extend_storefront.md:2]:
```suggestion
description: Extend Storefront with new menus
```

File: docs/commerce/storefront/extend_storefront.md [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+## Built-in menus
+
+With the `ibexa\storefront` package come the following built-in menus:
+
+| Item   | Value     | Description |
+|------------|----------|---------|
+| [Breadcrumbs](#breadcrumbs-menu)| - | Renders breadcrumbs for Content Tree Root, Taxonomy Entry, Product, User settings and User settings group | 
+| [Taxonomy](#taxonomy-menu)| - | Renders a menu to change the active currency |               
+| Currency| `currency_menu` | Renders a menu to change the active currency |               
+| Currency| `currency_menu` | Renders a menu to change the active currency |
```
Comment (DominikaK) [inline docs/commerce/storefront/extend_storefront.md:17]:
```suggestion
```

File: docs/commerce/storefront/extend_storefront.md [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+
+With the `ibexa\storefront` package come the following built-in menus:
+
+| Item   | Value     | Description |
+|------------|----------|---------|
+| [Breadcrumbs](#breadcrumbs-menu)| - | Renders breadcrumbs for Content Tree Root, Taxonomy Entry, Product, User settings and User settings group | 
+| [Taxonomy](#taxonomy-menu)| - | Renders a menu to change the active currency |               
+| Currency| `currency_menu` | Renders a menu to change the active currency |               
+| Currency| `currency_menu` | Renders a menu to change the active currency |               
+| Language| `langauge_menu` | Renders a menu to change the active language |
```
Comment (DominikaK) [inline docs/commerce/storefront/extend_storefront.md:18]:
```suggestion
| Language| `language_menu` | Renders a menu to change the active language |
```

File: docs/commerce/storefront/extend_storefront.md [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+| Region  | `region_menu`  | Renders a menu to change thr active region |
```
Comment (DominikaK) [inline docs/commerce/storefront/extend_storefront.md:19]:
```suggestion
| Region  | `region_menu`  | Renders a menu to change the active region |
```

File: docs/commerce/storefront/extend_storefront.md [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+{% set currrency_menu = knp_menu_get('ibexa_storefront.menu.currency') %}
+
+{{ knp_menu_render(currrency_menu) }}
+```
+
+### Breadcrumbs menu
+
+To add breadcrumbs menu to the product, you need to use an event subscriber.
+
+Create an event subscriber in the `src/EventSubscriber/BreadcrumbsMenuSubscriber.php`:
```
Comment (DominikaK) [inline docs/commerce/storefront/extend_storefront.md:33]:
```suggestion
Create an event subscriber in `src/EventSubscriber/BreadcrumbsMenuSubscriber.php`:
```

Comment (DominikaK) [reply docs/commerce/storefront/extend_storefront.md:33]: Actually, it's not that way. Too add the menu you need only the templates below. You need the subscriber to modify the items in the menu

File: docs/commerce/storefront/extend_storefront.md [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+[[= include_file('code_samples/front/shop/storefront/src/EventSubscriber/BreadcrumbsMenuSubscriber.php') =]]
+```
+
+Next, create the `templates/themes/storefront/storefront/knp_menu/breadcrumbs.html.twig` template:
+
+```html+twig
+[[= include_file('code_samples/front/shop/storefront/templates/themes/storefront/storefront/knp_menu/breadcrumbs.html.twig') =]]
+```
+
+Next, extend the `templates/themes/storefront/storefront/product.html.twig` template:
```
Comment (DominikaK) [inline docs/commerce/storefront/extend_storefront.md:45]:
```suggestion
Next, extend the `templates/themes/storefront/storefront/product.html.twig` template to include the breadcrumbs:
```

File: docs/commerce/storefront/storefront.md [hunk 21]
```diff
@@ -0,0 +1,28 @@
+---
+description: Storefront covers all actions related to the purchase process.
+edition: commerce
+---
+
+# Storefront
+
+Storefront package provides a starting kit for the developers.
```
Comment (DominikaK) [inline docs/commerce/storefront/storefront.md:8]:
```suggestion
The Storefront package provides a starting kit for the developers.
```

File: docs/commerce/storefront/storefront.md [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+| Search for specific product | Allows for searching for products, for example on the product listing page |
+| Sort products | Enables sorting products based on different criteria on a product listing page |
+| Buy button | Enables adding products to cart on a product listing and product details page |
+| Menus | Provide components that enable selecting different regions or currencies, for example to dynamically change the contents of the product listing page |
+| Login/register page |  Provides user interface for the login/registration page that enables buyers to access the Product catalog|
+
+
+!!!note
+
+    For information about roles and permissions that control access to various components of the purchase process, see [Permission use cases](../../permissions/permission_use_cases.md#commerce).
```
Comment (DominikaK) [inline docs/commerce/storefront/storefront.md:28]:
```suggestion
    For information about roles and permissions that control access to various components of the purchase process, see [Permission use cases](permission_use_cases.md#commerce).
```

**Review by DominikaK — CHANGES_REQUESTED** [review]

**Review by DominikaK — APPROVED** [review]

File: code_samples/front/shop/catalog_custom_logic/CatalogResolveEvent.php [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+
+    public function onCatalogResolve(CatalogResolveEvent $event): void
+    {
+        $user = $event->getUser();
+        if (str_ends_with($user->email, '@ibexa.co')) {
+            // Custom catalog for Ibexa employees 
+            $event->setCatalog($this->catalogService->getCatalogByIdentifier('employees'));
+        }      
+    }
+}
```
Comment (adamwojs) [inline code_samples/front/shop/catalog_custom_logic/CatalogResolveEvent.php:34]: New line is missing

File: docs/commerce/storefront/extend_storefront.md [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+You can build menu items based on repository objects with the `\Ibexa\Contracts\Storefront\Menu\ItemFactoryInterface` methods.
+
+The following items are available:
+
+- Content
+- Content ID
+- Location
+- Location ID
+- Taxonomy Entry
+- Product
```
Comment (adamwojs) [inline docs/commerce/storefront/extend_storefront.md:91]:
```suggestion
## Create menu items

`\Ibexa\Contracts\Storefront\Menu\ItemFactoryInterface` provides convenient methods to build menu item based on repository objects, including:

- Content
- Content ID
- Location
- Location ID
- Taxonomy Entry
- Product
```

In current shape this section mixes two unrelated topics

**Review by adamwojs — COMMENTED** [review]
It's good starting point but we need cover all storefront features .

**Comment by juskora** [comment]
@adamwojs So far it's a MVP but we will gradually add Storefront features from the scope.

**Review by DominikaK — APPROVED** [review]

File: code_samples/front/shop/storefront/templates/themes/storefront/storefront/product.html.twig [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+    {% set breadcrumbs_menu = knp_menu_get('ibexa_storefront.menu.breadcrumbs.product', [], {
+        'product': product
+    }) %}
+    {{ knp_menu_render(breadcrumbs_menu, {
+        template: '@ibexadesign/storefront/knp_menu/breadcrumbs.html.twig',
+    }) }}
+    {% include '@ibexadesign/storefront/component/product_view.html.twig' %}
+
+
+{% endblock %}
```
Comment (adamwojs) [inline code_samples/front/shop/storefront/templates/themes/storefront/storefront/product.html.twig:15]:
```suggestion
{% block content %}
    {% set breadcrumbs_menu = knp_menu_get('ibexa_storefront.menu.breadcrumbs.product', [], {
        'product': product        
    }) %}
    
    {{ knp_menu_render(breadcrumbs_menu, {
        template: '@ibexadesign/storefront/knp_menu/breadcrumbs.html.twig',
    }) }}
    
    {% include '@ibexadesign/storefront/component/product_view.html.twig' %}
{% endblock %}
```

File: docs/commerce/storefront/configure_storefront.md [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+namespace Ibexa\Contracts\Storefront\Repository;
+
+use Ibexa\Contracts\Core\Repository\Values\User\User;
+use Ibexa\Contracts\ProductCatalog\Values\CatalogInterface;
+
+interface CatalogResolverInterface
+{
+    public function resolveCatalog(?User $user = null): ?CatalogInterface;
+}
+```
```
Comment (adamwojs) [inline docs/commerce/storefront/configure_storefront.md:81]:
```suggestion
```php
namespace Ibexa\Contracts\Storefront\Repository;

use Ibexa\Contracts\Core\Repository\Values\User\User;
use Ibexa\Contracts\ProductCatalog\Values\CatalogInterface;

interface CatalogResolverInterface
{
    public function resolveCatalog(?User $user = null): ?CatalogInterface;
}
```
```

This should fix a syntax highlighting on GH

**Review by adamwojs — APPROVED** [review]

**Pull request merged**